File: runner/ts/src/wireRunner.ts
// Created by Yanjunhui

import { MongoClient, Db, Collection } from 'mongodb';
import { Database, ProtocolServer } from 'monolite';
import { TestCase, TestResult } from './types';
import { deserialize } from 'bson';

/**
 * Wire Protocol 测试运行器类
 * 通过 MongoDB Wire Protocol 连接 MonoLite 数据库执行测试
 * // EN: Wire Protocol test runner class
 * // EN: Executes tests by connecting to MonoLite database via MongoDB Wire Protocol
 */
export class WireRunner {
  /** MonoLite 数据库实例 // EN: MonoLite database instance */
  private db: Database | null = null;
  /** Wire Protocol 服务器实例 // EN: Wire Protocol server instance */
  private server: ProtocolServer | null = null;
  /** MongoDB 客户端实例 // EN: MongoDB client instance */
  private client: MongoClient | null = null;
  /** MongoDB 数据库实例 // EN: MongoDB database instance */
  private mongoDb: Db | null = null;
  /** 数据库文件路径 // EN: Database file path */
  private dbPath: string;
  /** Wire Protocol 端口号 // EN: Wire Protocol port number */
  private port: number;

  /**
   * 构造函数
   * // EN: Constructor
   * @param dbPath 数据库文件路径 // EN: Database file path
   * @param port Wire Protocol 端口号，默认 27019 // EN: Wire Protocol port, default 27019
   */
  constructor(dbPath: string, port: number = 27019) {
    this.dbPath = dbPath;
    this.port = port;
  }

  /**
   * 打开数据库连接并启动 Wire Protocol 服务器
   * // EN: Open database connection and start Wire Protocol server
   */
  async open(): Promise<void> {
    // 打开 MonoLite 数据库 // EN: Open MonoLite database
    this.db = await Database.open({ filePath: this.dbPath });

    // 启动 Wire Protocol 服务器 // EN: Start Wire Protocol server
    this.server = new ProtocolServer(`localhost:${this.port}`, this.db);
    await this.server.start();

    // 等待服务器启动 // EN: Wait for server to start
    await new Promise((resolve) => setTimeout(resolve, 100));

    // 连接 MongoDB 客户端 // EN: Connect MongoDB client
    this.client = new MongoClient(`mongodb://localhost:${this.port}`, {
      directConnection: true,
    });
    await this.client.connect();
    this.mongoDb = this.client.db('test');
  }

  /**
   * 关闭所有连接和服务器
   * // EN: Close all connections and server
   */
  async close(): Promise<void> {
    if (this.client) {
      await this.client.close();
      this.client = null;
    }
    if (this.server) {
      await this.server.stop();
      this.server = null;
    }
    if (this.db) {
      await this.db.close();
      this.db = null;
    }
  }

  /**
   * 运行单个测试用例
   * // EN: Run a single test case
   * @param tc 测试用例 // EN: Test case
   * @returns 测试结果 // EN: Test result
   */
  async runTest(tc: TestCase): Promise<TestResult> {
    const start = Date.now();
    const result: TestResult = {
      test_name: tc.name,
      language: 'typescript',
      mode: 'wire',
      success: false,
      duration_ms: 0,
    };

    try {
      if (!this.mongoDb) throw new Error('MongoDB not connected');

      const col = this.mongoDb.collection(tc.collection);

      // 执行前置步骤 // EN: Execute setup steps
      await this.executeSetup(col, tc);

      // 执行测试动作 // EN: Execute test action
      await this.executeAction(col, tc, result);

      result.success = true;
    } catch (err: any) {
      const errorMsg = err.message || String(err);
      result.error = errorMsg;
      // 检查是否是预期的错误 // EN: Check if it's an expected error
      if (tc.expected.error && errorMsg.includes(tc.expected.error)) {
        result.success = true;
      }
    }

    result.duration_ms = Date.now() - start;
    return result;
  }

  /**
   * 执行测试前置步骤
   * // EN: Execute test setup steps
   * @param col MongoDB 集合 // EN: MongoDB collection
   * @param tc 测试用例 // EN: Test case
   */
  private async executeSetup(col: Collection, tc: TestCase): Promise<void> {
    if (!tc.setup || tc.setup.length === 0) return;

    for (const step of tc.setup) {
      const data = this.parseData(step.data);
      switch (step.operation) {
        case 'insert':
          // 插入单个文档 // EN: Insert single document
          await col.insertOne(data);
          break;
        case 'createIndex':
          // 创建索引 // EN: Create index
          const keys = data.keys || {};
          const opts = data.options || {};
          await col.createIndex(keys, opts);
          break;
      }
    }
  }

  /**
   * 执行测试动作
   * // EN: Execute test action
   * @param col MongoDB 集合 // EN: MongoDB collection
   * @param tc 测试用例 // EN: Test case
   * @param result 测试结果对象 // EN: Test result object
   */
  private async executeAction(
    col: Collection,
    tc: TestCase,
    result: TestResult
  ): Promise<void> {
    const action = tc.action;
    // 解析各种参数 // EN: Parse various parameters
    const filter = action.filter ? this.parseData(action.filter) : {};
    const update = action.update ? this.parseData(action.update) : {};
    const doc = action.doc ? this.parseData(action.doc) : {};
    const docs = action.docs ? this.parseData(action.docs) : [];
    const options = action.options ? this.parseData(action.options) : {};

    switch (action.method) {
      case 'insertOne': {
        // 插入单个文档 // EN: Insert single document
        await col.insertOne(doc);
        result.count = 1;
        break;
      }

      case 'insertMany': {
        // 批量插入文档 // EN: Insert multiple documents
        const res = await col.insertMany(docs);
        result.count = res.insertedCount;
        break;
      }

      case 'find': {
        // 查询文档 // EN: Find documents
        let cursor = col.find(filter);
        if (options.sort) cursor = cursor.sort(options.sort);
        if (options.skip) cursor = cursor.skip(options.skip);
        if (options.limit) cursor = cursor.limit(options.limit);
        if (options.projection) cursor = cursor.project(options.projection);

        const results = await cursor.toArray();
        result.count = results.length;
        result.documents = results;
        break;
      }

      case 'findOne': {
        // 查询单个文档 // EN: Find single document
        const doc = await col.findOne(filter);
        if (doc) {
          result.count = 1;
          result.documents = [doc];
        } else {
          result.count = 0;
        }
        break;
      }

      case 'updateOne': {
        // 更新单个文档 // EN: Update single document
        const upsert = options.upsert || false;
        const res = await col.updateOne(filter, update, { upsert });
        result.matched_count = res.matchedCount;
        result.modified_count = res.modifiedCount;
        result.upserted_id = res.upsertedId;
        break;
      }

      case 'updateMany': {
        // 批量更新文档 // EN: Update multiple documents
        const res = await col.updateMany(filter, update);
        result.matched_count = res.matchedCount;
        result.modified_count = res.modifiedCount;
        break;
      }

      case 'deleteOne': {
        // 删除单个文档 // EN: Delete single document
        const res = await col.deleteOne(filter);
        result.deleted_count = res.deletedCount;
        break;
      }

      case 'deleteMany': {
        // 批量删除文档 // EN: Delete multiple documents
        const res = await col.deleteMany(filter);
        result.deleted_count = res.deletedCount;
        break;
      }

      case 'replaceOne': {
        // 替换单个文档 // EN: Replace single document
        const res = await col.replaceOne(filter, doc);
        result.matched_count = res.matchedCount;
        result.modified_count = res.modifiedCount;
        break;
      }

      case 'distinct': {
        // 获取字段的去重值 // EN: Get distinct values of a field
        const field = options.field || '';
        const values = await col.distinct(field, filter);
        result.count = values.length;
        break;
      }

      case 'aggregate': {
        // 聚合查询 // EN: Aggregate query
        const pipeline = options.pipeline || [];
        const cursor = col.aggregate(pipeline);
        const docs = await cursor.toArray();
        result.count = docs.length;
        result.documents = docs;
        break;
      }

      case 'createIndex': {
        // 创建索引 // EN: Create index
        const keys = options.keys || {};
        const indexOpts = options.options || {};
        const name = await col.createIndex(keys, indexOpts);
        result.count = 1;
        result.documents = [{ indexName: name }];
        break;
      }

      case 'listIndexes': {
        // 列出索引 // EN: List indexes
        const indexes = await col.listIndexes().toArray();
        result.count = indexes.length;
        break;
      }

      case 'dropIndex': {
        // 删除索引 // EN: Drop index
        const name = options.name || '';
        await col.dropIndex(name);
        break;
      }

      default:
        // 未知方法 // EN: Unknown method
        throw new Error(`Unknown method: ${action.method}`);
    }
  }

  /**
   * 解析数据，处理 BSON 序列化格式
   * // EN: Parse data, handle BSON serialization format
   * @param data 原始数据 // EN: Raw data
   * @returns 解析后的数据 // EN: Parsed data
   */
  private parseData(data: any): any {
    if (data === null || data === undefined) return data;

    // 如果是 Buffer 或 Uint8Array（BSON 原始格式）
    // // EN: If it's Buffer or Uint8Array (BSON raw format)
    if (Buffer.isBuffer(data) || data instanceof Uint8Array) {
      return deserialize(Buffer.from(data));
    }

    // 如果是对象，检查是否有 $binary 格式（JSON 序列化的 BSON）
    // // EN: If it's an object, check for $binary format (JSON serialized BSON)
    if (typeof data === 'object') {
      if (data.$binary && data.$binary.base64) {
        const buf = Buffer.from(data.$binary.base64, 'base64');
        return deserialize(buf);
      }
      return data;
    }

    return data;
  }
}

File: runner/go/wire_runner.go
// Created by Yanjunhui

package main

import (
	"context"
	"fmt"
	"time"

	"github.com/monolite/monodb/engine"
	"github.com/monolite/monodb/protocol"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// WireRunner 通过 Wire Protocol 测试
// EN: WireRunner tests through Wire Protocol.
type WireRunner struct {
	db     *engine.Database   // 数据库实例 // EN: Database instance
	server *protocol.Server   // Wire Protocol 服务器 // EN: Wire Protocol server
	client *mongo.Client      // MongoDB 客户端 // EN: MongoDB client
	addr   string             // 服务器地址 // EN: Server address
}

// NewWireRunner 创建 Wire 运行器
// EN: NewWireRunner creates a Wire runner.
func NewWireRunner(dbPath string, port int) (*WireRunner, error) {
	db, err := engine.OpenDatabase(dbPath)
	if err != nil {
		return nil, fmt.Errorf("打开数据库失败: %w", err) // EN: Failed to open database
	}

	addr := fmt.Sprintf(":%d", port)
	server := protocol.NewServer(addr, db)

	// 启动服务器 // EN: Start server
	go func() {
		if err := server.Start(); err != nil {
			fmt.Printf("服务器启动失败: %v\n", err) // EN: Server start failed
		}
	}()

	// 等待服务器启动 // EN: Wait for server to start
	time.Sleep(100 * time.Millisecond)

	// 连接客户端 // EN: Connect client
	ctx := context.Background()
	clientOpts := options.Client().
		ApplyURI(fmt.Sprintf("mongodb://localhost:%d", port)).
		SetDirect(true)

	client, err := mongo.Connect(ctx, clientOpts)
	if err != nil {
		db.Close()
		return nil, fmt.Errorf("连接客户端失败: %w", err) // EN: Failed to connect client
	}

	return &WireRunner{
		db:     db,
		server: server,
		client: client,
		addr:   addr,
	}, nil
}

// Close 关闭连接
// EN: Close closes all connections.
func (r *WireRunner) Close() error {
	ctx := context.Background()
	if r.client != nil {
		r.client.Disconnect(ctx)
	}
	if r.server != nil {
		r.server.Stop()
	}
	if r.db != nil {
		return r.db.Close()
	}
	return nil
}

// RunTest 运行单个测试
// EN: RunTest runs a single test case.
func (r *WireRunner) RunTest(tc TestCase) TestResult {
	start := time.Now()
	result := TestResult{
		TestName: tc.Name,
		Language: "go",
		Mode:     "wire",
	}

	ctx := context.Background()
	col := r.client.Database("test").Collection(tc.Collection)

	// 执行前置步骤 // EN: Execute setup steps
	if err := r.executeSetup(ctx, col, tc); err != nil {
		result.Error = fmt.Sprintf("Setup 失败: %v", err) // EN: Setup failed
		result.Duration = time.Since(start).Milliseconds()
		return result
	}

	// 执行测试动作 // EN: Execute test action
	if err := r.executeAction(ctx, col, tc, &result); err != nil {
		result.Error = err.Error()
		result.Duration = time.Since(start).Milliseconds()
		if tc.Expected.Error != "" {
			result.Success = true
		}
		return result
	}

	result.Success = true
	result.Duration = time.Since(start).Milliseconds()
	return result
}

// executeSetup 执行前置步骤
// EN: executeSetup executes the setup steps.
func (r *WireRunner) executeSetup(ctx context.Context, col *mongo.Collection, tc TestCase) error {
	if len(tc.Setup) == 0 {
		return nil
	}

	for _, step := range tc.Setup {
		switch step.Operation {
		case "insert":
			doc := toBsonD(step.Data)
			if _, err := col.InsertOne(ctx, doc); err != nil {
				return err
			}
		case "createIndex":
			opts := toBsonD(step.Data)
			keysRaw := getField(opts, "keys")
			keys := toBsonD(keysRaw)
			indexModel := mongo.IndexModel{Keys: keys}
			if indexOpts := getFieldD(opts, "options"); indexOpts != nil {
				if v := getField(indexOpts, "unique"); v != nil {
					unique := v.(bool)
					indexModel.Options = options.Index().SetUnique(unique)
				}
				if v := getField(indexOpts, "name"); v != nil {
					if indexModel.Options == nil {
						indexModel.Options = options.Index()
					}
					indexModel.Options.SetName(v.(string))
				}
			}
			if _, err := col.Indexes().CreateOne(ctx, indexModel); err != nil {
				return err
			}
		}
	}
	return nil
}

// executeAction 执行测试动作
// EN: executeAction executes the test action.
func (r *WireRunner) executeAction(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	switch tc.Action.Method {
	case "insertOne":
		return r.executeInsertOne(ctx, col, tc, result)
	case "insertMany":
		return r.executeInsertMany(ctx, col, tc, result)
	case "find":
		return r.executeFind(ctx, col, tc, result)
	case "findOne":
		return r.executeFindOne(ctx, col, tc, result)
	case "updateOne":
		return r.executeUpdateOne(ctx, col, tc, result)
	case "updateMany":
		return r.executeUpdateMany(ctx, col, tc, result)
	case "deleteOne":
		return r.executeDeleteOne(ctx, col, tc, result)
	case "deleteMany":
		return r.executeDeleteMany(ctx, col, tc, result)
	case "replaceOne":
		return r.executeReplaceOne(ctx, col, tc, result)
	case "aggregate":
		return r.executeAggregate(ctx, col, tc, result)
	case "distinct":
		return r.executeDistinct(ctx, col, tc, result)
	case "createIndex":
		return r.executeCreateIndex(ctx, col, tc, result)
	case "listIndexes":
		return r.executeListIndexes(ctx, col, tc, result)
	case "dropIndex":
		return r.executeDropIndex(ctx, col, tc, result)
	default:
		return fmt.Errorf("未知方法: %s", tc.Action.Method) // EN: Unknown method
	}
}

// executeInsertOne 执行插入单个文档
// EN: executeInsertOne executes insert one document.
func (r *WireRunner) executeInsertOne(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	doc := toBsonD(tc.Action.Doc)
	_, err := col.InsertOne(ctx, doc)
	if err != nil {
		return err
	}
	result.Count = 1
	return nil
}

// executeInsertMany 执行批量插入文档
// EN: executeInsertMany executes insert many documents.
func (r *WireRunner) executeInsertMany(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	docs := toBsonDSlice(tc.Action.Docs)
	ifaces := make([]interface{}, len(docs))
	for i, d := range docs {
		ifaces[i] = d
	}
	res, err := col.InsertMany(ctx, ifaces)
	if err != nil {
		return err
	}
	result.Count = int64(len(res.InsertedIDs))
	return nil
}

// executeFind 执行查询
// EN: executeFind executes a find query.
func (r *WireRunner) executeFind(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	var filter bson.D
	if tc.Action.Filter != nil {
		filter = toBsonD(tc.Action.Filter)
	}

	findOpts := options.Find()
	if tc.Action.Options != nil {
		opts := toBsonD(tc.Action.Options)
		if v := getField(opts, "sort"); v != nil {
			findOpts.SetSort(toBsonD(v))
		}
		if v := getField(opts, "limit"); v != nil {
			findOpts.SetLimit(toInt64(v))
		}
		if v := getField(opts, "skip"); v != nil {
			findOpts.SetSkip(toInt64(v))
		}
		if v := getField(opts, "projection"); v != nil {
			findOpts.SetProjection(toBsonD(v))
		}
	}

	cursor, err := col.Find(ctx, filter, findOpts)
	if err != nil {
		return err
	}
	defer cursor.Close(ctx)

	var docs []bson.M
	if err := cursor.All(ctx, &docs); err != nil {
		return err
	}
	result.Count = int64(len(docs))
	result.Documents = docs
	return nil
}

// executeFindOne 执行查询单个文档
// EN: executeFindOne executes a find one query.
func (r *WireRunner) executeFindOne(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	filter := toBsonD(tc.Action.Filter)

	var doc bson.M
	err := col.FindOne(ctx, filter).Decode(&doc)
	if err == mongo.ErrNoDocuments {
		result.Count = 0
		return nil
	}
	if err != nil {
		return err
	}
	result.Count = 1
	result.Documents = []bson.M{doc}
	return nil
}

// executeUpdateOne 执行更新单个文档
// EN: executeUpdateOne executes update one document.
func (r *WireRunner) executeUpdateOne(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	filter := toBsonD(tc.Action.Filter)
	update := toBsonD(tc.Action.Update)

	updateOpts := options.Update()
	if tc.Action.Options != nil {
		opts := toBsonD(tc.Action.Options)
		if v := getField(opts, "upsert"); v != nil {
			updateOpts.SetUpsert(v.(bool))
		}
	}

	res, err := col.UpdateOne(ctx, filter, update, updateOpts)
	if err != nil {
		return err
	}
	result.MatchedCount = res.MatchedCount
	result.ModifiedCount = res.ModifiedCount
	result.UpsertedID = res.UpsertedID
	return nil
}

// executeUpdateMany 执行更新多个文档
// EN: executeUpdateMany executes update many documents.
func (r *WireRunner) executeUpdateMany(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	filter := toBsonD(tc.Action.Filter)
	update := toBsonD(tc.Action.Update)

	res, err := col.UpdateMany(ctx, filter, update)
	if err != nil {
		return err
	}
	result.MatchedCount = res.MatchedCount
	result.ModifiedCount = res.ModifiedCount
	return nil
}

// executeDeleteOne 执行删除单个文档
// EN: executeDeleteOne executes delete one document.
func (r *WireRunner) executeDeleteOne(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	filter := toBsonD(tc.Action.Filter)
	res, err := col.DeleteOne(ctx, filter)
	if err != nil {
		return err
	}
	result.DeletedCount = res.DeletedCount
	return nil
}

// executeDeleteMany 执行删除多个文档
// EN: executeDeleteMany executes delete many documents.
func (r *WireRunner) executeDeleteMany(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	filter := toBsonD(tc.Action.Filter)
	res, err := col.DeleteMany(ctx, filter)
	if err != nil {
		return err
	}
	result.DeletedCount = res.DeletedCount
	return nil
}

// executeReplaceOne 执行替换单个文档
// EN: executeReplaceOne executes replace one document.
func (r *WireRunner) executeReplaceOne(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	filter := toBsonD(tc.Action.Filter)
	replacement := toBsonD(tc.Action.Doc)
	res, err := col.ReplaceOne(ctx, filter, replacement)
	if err != nil {
		return err
	}
	result.MatchedCount = res.MatchedCount
	result.ModifiedCount = res.ModifiedCount
	return nil
}

// executeAggregate 执行聚合管道
// EN: executeAggregate executes aggregation pipeline.
func (r *WireRunner) executeAggregate(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	opts := toBsonD(tc.Action.Options)
	pipelineRaw := getField(opts, "pipeline")

	// 转换为 bson.A // EN: Convert to bson.A
	var pipeline bson.A
	switch p := pipelineRaw.(type) {
	case []interface{}:
		for _, stage := range p {
			pipeline = append(pipeline, toBsonD(stage))
		}
	case bson.A:
		for _, stage := range p {
			pipeline = append(pipeline, toBsonD(stage))
		}
	default:
		return fmt.Errorf("pipeline 类型错误: %T", pipelineRaw) // EN: Pipeline type error
	}

	cursor, err := col.Aggregate(ctx, pipeline)
	if err != nil {
		return err
	}
	defer cursor.Close(ctx)

	var docs []bson.M
	if err := cursor.All(ctx, &docs); err != nil {
		return err
	}
	result.Count = int64(len(docs))
	result.Documents = docs
	return nil
}

// executeDistinct 执行去重查询
// EN: executeDistinct executes distinct query.
func (r *WireRunner) executeDistinct(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	var filter bson.D
	if tc.Action.Filter != nil {
		filter = toBsonD(tc.Action.Filter)
	}

	var field string
	if tc.Action.Options != nil {
		opts := toBsonD(tc.Action.Options)
		if v := getField(opts, "field"); v != nil {
			field = v.(string)
		}
	}

	values, err := col.Distinct(ctx, field, filter)
	if err != nil {
		return err
	}
	result.Count = int64(len(values))
	return nil
}

// executeCreateIndex 执行创建索引
// EN: executeCreateIndex executes create index operation.
func (r *WireRunner) executeCreateIndex(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	opts := toBsonD(tc.Action.Options)

	keysRaw := getField(opts, "keys")
	keys := toBsonD(keysRaw)
	indexModel := mongo.IndexModel{Keys: keys}

	if indexOpts := getFieldD(opts, "options"); indexOpts != nil {
		indexModel.Options = options.Index()
		if v := getField(indexOpts, "unique"); v != nil {
			indexModel.Options.SetUnique(v.(bool))
		}
		if v := getField(indexOpts, "name"); v != nil {
			indexModel.Options.SetName(v.(string))
		}
	}

	name, err := col.Indexes().CreateOne(ctx, indexModel)
	if err != nil {
		return err
	}
	result.Count = 1
	result.Documents = []bson.M{{"indexName": name}}
	return nil
}

// executeListIndexes 执行列出索引
// EN: executeListIndexes executes list indexes operation.
func (r *WireRunner) executeListIndexes(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	cursor, err := col.Indexes().List(ctx)
	if err != nil {
		return err
	}
	defer cursor.Close(ctx)

	var indexes []bson.M
	if err := cursor.All(ctx, &indexes); err != nil {
		return err
	}
	result.Count = int64(len(indexes))
	return nil
}

// executeDropIndex 执行删除索引
// EN: executeDropIndex executes drop index operation.
func (r *WireRunner) executeDropIndex(ctx context.Context, col *mongo.Collection, tc TestCase, result *TestResult) error {
	opts := toBsonD(tc.Action.Options)
	name := getField(opts, "name").(string)
	_, err := col.Indexes().DropOne(ctx, name)
	return err
}
